### PR TITLE
Ag image tags

### LINKED ISSUE #7: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by almgong**
Closes #7

### CODE REVIEW

File: db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
Comment (shfung) [inline db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Maybe we can add some not-null constraints to some of the columns?

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -28,7 +28,9 @@ def show
   # POST /images
   def create
     # only keep parameters we expect
-    @image = Image.new image_params
+    params = image_params
+    params[:tag_list] = params[:tag_list]&.split&.join(',')
```
Comment (shfung) [inline app/controllers/images_controller.rb]: What will happen if the user wants to enter a tag with a comma in it? e.g. The user wants to add two tags: `a,b` and `c`.

File: test/controllers/image_controller_test.rb [hunk 3]
```diff
@@ -8,6 +8,13 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
     assert_select 'img[src=?]', 'http://example.com/picture1.jpg'
   end
 
+  test 'should GET valid image tags successfully' do
+    image = Image.create! url: 'http://example.com/picture1.jpg', tag_list: 'tag1,tag2'
+    get image_path image
```
Comment (shfung) [inline test/controllers/image_controller_test.rb]: I think it'll be easier to read with parenthesis when you have nested parameters.

File: test/controllers/image_controller_test.rb [hunk 4]
```diff
@@ -16,29 +23,24 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
...
-    valid_img_urls = [
-      'http://example.com/some-img.jpg',
-      'https://google.com/some-img.gif',
-      'http://no.protocol.com/img-dir/file.jpg'
-    ]
-    valid_img_urls.each { |url| post_valid_img_url(url) }
+    url = 'http://example.com/some-img.jpg'
+
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: url } }
```
Comment (shfung) [inline test/controllers/image_controller_test.rb]: Since you have added `tag_list` to your form, I think we should include the key `tag_list` to all the `post` requests to the create action.

Comment (shfung) [inline test/controllers/image_controller_test.rb]: Maybe add a test to make sure the `create` action takes a `tag_list` param and uses it correctly.

File: db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (almgong) [inline db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Looking through the schema for tables created by the gem, it seems like tags.name, taggings.tag_id, and taggings.taggable_id should be not nullable. If we ever associate taggers, then taggings.tagger_id would make sense too (but for now since we don't have user accounts, I will ignore this case). 

Still looking to see if taggings.taggable_type is worth making not-nullable, as I don't quite understand what it is used for yet.

File: test/controllers/image_controller_test.rb (hunk 4 above)
Comment (almgong) [inline test/controllers/image_controller_test.rb]: I have a test on line 11 ` test 'should GET valid image tags successfully'` that checks that the tags exist on the show page, for a given image and its tags. I think that since I have this check, adding another test with a check for the tags would add some redundancy. Perhaps I could simply check that the redirection above (e.g. `assert_redirected_to Image.last`) succeeded with a 200 code? 

This way by testing the show action separately, I also test the behavior after the redirection without having to duplicate code. What are your thoughts?

File: db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (shfung) [inline db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: The `_id` and `_type` pair is what is called polymorphism association; it allows you to associate the model with multiple different tables.  e.g., you can use the same tag table to tag `Image` or say `Comment`, then the `taggable_type` will be storing either `Image` or `Comment`.

Comment (almgong) [inline db/migrate/20170316201955_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Oh, thanks for clearing that up. `_type` should be not nullable as well since it seems like an important association. I will make these fields not nullable in the generated migration files.

File: app/views/images/index.html.erb [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
+    <% @images.each do |image| %>
+      <div class="col-lg-4 col-sm-6 image-card">
+        <div class="card">
+          <%= image_tag(image.url, alt: 'User uploaded image', class: 'card-img-top') %>
+          <div class="card-block">
+            <p class="card-text">
+              <% if image.tag_list.empty? %>
+                <i>No tags</i>
+              <% end %>
+              <%= image.tag_list.join ' ' %>
```
Comment (shfung) [inline app/views/images/index.html.erb]: Now that the tags are comma separated, I'd suggest using comma to separate them when displaying.

File: app/views/images/show.html.erb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 <div class="text-center mt-4">
   <h2>Visit original <a href="<%= @image.url %>" target="_blank">here</a></h2>
+  <h3><%= @image.tag_list.join ' ' %></h3>
```
Comment (shfung) [inline app/views/images/show.html.erb]: Now that the tags are comma separated, I'd suggest using comma to separate them when displaying.

File: app/views/images/new.html.erb [hunk 7]
```diff
@@ -6,6 +6,9 @@
           <%= form.input :url, label: 'Enter your image url here:', as: :string, input_html: { class: 'form-control' } %>
         </div>
         <div class="form-group">
+          <%= form.input :tag_list, label: 'Tags for your image (comma separated)', input_html: { class: 'form-control' } %>
```
Comment (shfung) [inline app/views/images/new.html.erb]: If you submit the form with an invalid URL and comma separated tags, are the tags still comma separated after you got back an error?

File: db/schema.rb [hunk 8]
```diff
@@ -19,4 +19,29 @@
...
+    t.index ["taggable_id", "taggable_type", "tagger_id", "context"], name: "taggings_idy"
+    t.index ["taggable_id"], name: "index_taggings_on_taggable_id"
+    t.index ["taggable_type"], name: "index_taggings_on_taggable_type"
+    t.index ["tagger_id", "tagger_type"], name: "index_taggings_on_tagger_id_and_tagger_type"
+    t.index ["tagger_id"], name: "index_taggings_on_tagger_id"
+  end
+
+  create_table "tags", force: :cascade do |t|
+    t.string  "name",                       null: false
+    t.integer "taggings_count", default: 0
```
Comment (shfung) [inline db/schema.rb]: I think `context`, `created_at` and `taggings_count` can have not-null constraints too.

File: test/controllers/image_controller_test.rb [hunk 9]
```diff
@@ -16,30 +23,25 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
...
-    valid_img_urls = [
-      'http://example.com/some-img.jpg',
-      'https://google.com/some-img.gif',
-      'http://no.protocol.com/img-dir/file.jpg'
-    ]
-    valid_img_urls.each { |url| post_valid_img_url(url) }
+    url = 'http://example.com/some-img.jpg'
+
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: url, tag_list: nil } }
```
Comment (shfung) [inline test/controllers/image_controller_test.rb]: Maybe add a test to post with tag?

File: app/views/images/new.html.erb (hunk 7 above)
Comment (almgong) [inline app/views/images/new.html.erb]: It seems like the commas disappear. Doing some research led to: https://github.com/mbleigh/acts-as-taggable-on/issues/620. So to my understanding it is because the acts_as_taggable_on gem does not invoke html_safe on the tag_list value by default. I will add a value attribute with a call to `.to_s` to address this problem.

File: test/controllers/image_controller_test.rb [hunk 10]
```diff
@@ -16,30 +23,36 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
...
 
-      post images_path, params: { image: { url: 'not-a-url' } }
-      assert_response :unprocessable_entity
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: url, tag_list: tags } }
+    end
 
-      post images_path, params: { image: { url: 'http:/missing-a-forward-slash.com/img1.png' } }
-      assert_response :unprocessable_entity
+    assert_equal tags, Image.last.tag_list.to_s
```
Comment (shfung) [inline test/controllers/image_controller_test.rb]: One last comment: you should always assert on the response of the action of any controller test, i.e. you should add `assert_redirected_to Image.last` in this test.

Comment (almgong) [inline test/controllers/image_controller_test.rb]: Done!
